sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/DefaultSource.scala

```diff
@@ -91,6 +96,70 @@ class DefaultSource extends FileFormat with DataSourceRegister {
     new CSVOutputWriterFactory(csvOptions)
   }
 
+  override def buildReader(
+      sqlContext: SQLContext,
+      physicalSchema: StructType,
+      partitionSchema: StructType,
+      dataSchema: StructType,
+      filters: Seq[Filter],
+      options: Map[String, String]): (PartitionedFile) => Iterator[InternalRow] = {
+    val csvOptions = new CSVOptions(options)
+    val headers = dataSchema.fields.map(_.name)
+
+    val conf = new Configuration(sqlContext.sparkContext.hadoopConfiguration)
+    val broadcastedConf = sqlContext.sparkContext.broadcast(new SerializableConfiguration(conf))
+
+    (file: PartitionedFile) => {
+      val fileSplit = {
+        val filePath = new Path(new URI(file.filePath))
+        new FileSplit(filePath, file.start, file.length, Array.empty)
+      }
+
+      val hadoopAttemptContext = {
+        val conf = broadcastedConf.value.value
+        val attemptID = new TaskAttemptID(new TaskID(new JobID(), TaskType.MAP, 0), 0)
+        new TaskAttemptContextImpl(conf, attemptID)
+      }
+
+      val reader = new LineRecordReader()
+      reader.initialize(fileSplit, hadoopAttemptContext)
+
+      val lineIterator = new RecordReaderIterator(reader).map { line =>
+        new String(line.getBytes, 0, line.getLength, csvOptions.charset)
+      }
+
+      // Skips the header line of each file if the `header` option is set to true.
+      // TODO What if the first partitioned file consists of only comments and empty lines?
```

Comment:
Unlike RDD, here we no longer have a global view of all the input data. Thus filtering the first header line becomes difficult. Here we only handle the simplest and most common use case, namely we can find a valid header in the first block of any given input file. If the first block contains only comments and empty lines, this branch fails. Haven't got a good solution for this yet.